internal/postgres/report.go

```diff
@@ -0,0 +1,496 @@
+	pgSchema := conv.pgSchema[pgTable]
+	m := make(map[string][]schemaIssue)
+	warnings := int64(0)
+	warningBatcher := make(map[schemaIssue]bool)
+	// Note on how we count warnings when there are multiple warnings
+	// per column and/or multiple warnings per table.
+	// non-batched warnings: count at most one warning per column.
+	// batched warnings: count at most one warning per table.
+	for c, pc := range pgSchema.cols {
+		colWarning := false
+		m[c] = pc.issues
+		for _, i := range pc.issues {
+			switch {
+			case issueDB[i].severity == warning && issueDB[i].batch:
+				warningBatcher[i] = true
+			case issueDB[i].severity == warning && !issueDB[i].batch:
+				colWarning = true
+			}
+		}
+		if colWarning {
+			warnings++
+		}
+	}
+	warnings += int64(len(warningBatcher))
+	return m, int64(len(pgSchema.cols)), warnings
+}
+
+// rateSchema returns an string summarizing the quality of PostgreSQL
+// to Spanner schema conversion. 'cols' and 'warnings' are respectively
+// the number of columns converted and the warnings encountered
+// (both weighted by number of data rows).
+// 'missingPKey' indicates whether the PostgreSQL schema had a primary key.
+// 'summary' indicates whether this is a per-table rating or an overall
+// summary rating.
+func rateSchema(cols, warnings int64, missingPKey, summary bool) string {
+	pkMsg := "missing primary key"
+	if summary {
+		pkMsg = "some missing primary keys"
+	}
+	switch {
+	case cols == 0:
+		return "NONE (no schema found)"
+	case warnings == 0 && !missingPKey:
+		return "EXCELLENT (all columns mapped cleanly)"
+	case warnings == 0 && missingPKey:
+		return fmt.Sprintf("GOOD (all columns mapped cleanly, but %s)", pkMsg)
+	case good(cols, warnings) && !missingPKey:
+		return "GOOD (most columns mapped cleanly)"
+	case good(cols, warnings) && missingPKey:
+		return fmt.Sprintf("GOOD (most columns mapped cleanly, but %s)", pkMsg)
+	case ok(cols, warnings) && !missingPKey:
+		return "OK (some columns did not map cleanly)"
+	case ok(cols, warnings) && missingPKey:
+		return fmt.Sprintf("OK (some columns did not map cleanly + %s)", pkMsg)
+	case !missingPKey:
+		return "POOR (many columns did not map cleanly)"
+	default:
+		return fmt.Sprintf("POOR (many columns did not map cleanly + %s)", pkMsg)
+	}
+}
+
+func rateData(rows int64, badRows int64) string {
+	s := fmt.Sprintf(" (%s%% of %d rows written to Spanner)", pct(rows, badRows), rows)
+	switch {
+	case rows == 0:
+		return "NONE (no data rows found)"
+	case badRows == 0:
+		return fmt.Sprintf("EXCELLENT (all %d rows written to Spanner)", rows)
+	case good(rows, badRows):
+		return "GOOD" + s
+	case ok(rows, badRows):
+		return "OK" + s
+	default:
+		return "POOR" + s
+	}
+}
+
+func good(total, badCount int64) bool {
+	return badCount < total/20
+}
+
+func ok(total, badCount int64) bool {
+	return badCount < total/3
+}
+
+func prConvStats(rows, badRows, cols, warnings int64, missingPKey, summary bool) string {
```

Comment:
Rename to printConvStats or maybe sprintConvStats since it returns the string.

Reply:
How about: rateConversion?
Since it combines rateSchema and rateData to give overall rating info.